feat: [FC-0047] Extend mobile API with course progress and primary courses on dashboard view

KyryloKireiev opened the pull request:
<!--

🌳🌳
🌳🌳🌳🌳         🌳 Note: Quince is in support. Fixes you make on master may still be needed on Quince.
    🌳🌳🌳🌳     If so, make another pull request against the open-release/quince.master branch,
🌳🌳🌳🌳         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌳🌳


Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

Two APIs was updated: `UserCourseEnrollmentsList` and `BlocksInfoInCourseView`:

1. `UserCourseEnrollmentsList` - In v4 we added to the response primary object. Primary object contains the latest user's enrollment or course where user has the latest progress. Primary object has been cut from user's enrolments array and inserted into separated section with key `primary`.
Also we add to Primary enrollment object additional fields:
        * course_progress: Contains information about how many assignments are in the course
          and how many assignments the student has completed.
        * total_assignments_count: Total course's assignments count.
        * assignments_completed: Assignments witch the student has completed.
We added `?requested_fields=course_progress` query param to GET request:
`GET /api/mobile/v4/users/kyrylo/course_enrollments/?requested_fields=course_progress` With using this query param we can get `course_progress` field for the each student enrollment. 
Also, for the 4th version of this API, we changed the pagination - now the API returns 5 elements per page. This is necessary to improve performance.
Response example:
```
{
    "configs": {
        "iap_configs": {}
    },
    "user_timezone": "Europe/Kiev",
    "enrollments": {
        "next": null,
        "previous": null,
        "count": 2,
        "num_pages": 1,
        "current_page": 1,
        "start": 0,
        "results": [
            {
                "audit_access_expires": null,
                "created": "2024-05-06T15:48:23.943185Z",
                "mode": "audit",
                "is_active": true,
                "course": {
                    "id": "course-v1:assignments+assignments+assignments",
                    "name": "assignments",
                    "number": "assignments",
                    "org": "assignments",
                    "start": "2020-01-01T00:00:00Z",
                    "start_display": "Jan. 1, 2020",
                    "start_type": "timestamp",
                    "end": null,
                    "dynamic_upgrade_deadline": null,
                    "subscription_id": "course_MNXXK4TTMUWXMMJ2MFZXG2LHNZWWK3TUOMVWC43TNFTW43LFNZ2HGK3BONZWSZ3ONVSW45DT",
                    "courseware_access": {
                        "has_access": true,
                        "error_code": null,
                        "developer_message": null,
                        "user_message": null,
                        "additional_context_user_message": null,
                        "user_fragment": null
                    },
                    "media": {
                        "course_image": {
                            "uri": "/asset-v1:assignments+assignments+assignments+type@asset+block@15581993_5643241.jpg",
                            "name": "Course Image"
                        }
                    },
                    "course_image": "/asset-v1:assignments+assignments+assignments+type@asset+block@15581993_5643241.jpg",
                    "course_about": "http://local.edly.io:8000/courses/course-v1:assignments+assignments+assignments/about",
                    "course_sharing_utm_parameters": {
                        "facebook": "utm_medium=social&utm_campaign=social-sharing-db&utm_source=facebook",
                        "twitter": "utm_medium=social&utm_campaign=social-sharing-db&utm_source=twitter"
                    },
                    "course_updates": "http://local.edly.io:8000/api/mobile/v4/course_info/course-v1:assignments+assignments+assignments/updates",
                    "course_handouts": "http://local.edly.io:8000/api/mobile/v4/course_info/course-v1:assignments+assignments+assignments/handouts",
                    "discussion_url": null,
                    "video_outline": null,
                    "is_self_paced": false
                },
                "certificate": {},
                "course_modes": [
                    {
                        "slug": "audit",
                        "sku": null,
                        "android_sku": null,
                        "ios_sku": null,
                        "min_price": 0
                    }
                ],
                "course_progress": {
                    "total_assignments_count": 7,
                    "assignments_completed": 4
                }
            },
            {
                "audit_access_expires": null,
                "created": "2024-05-06T14:49:25.211198Z",
                "mode": "audit",
                "is_active": true,
                "course": {
                    "id": "course-v1:a+a+a",
                    "name": "a",
                    "number": "a",
                    "org": "a",
                    "start": "2030-01-01T00:00:00Z",
                    "start_display": null,
                    "start_type": "empty",
                    "end": null,
                    "dynamic_upgrade_deadline": null,
                    "subscription_id": "course_MNXXK4TTMUWXMMJ2MEVWCK3B",
                    "courseware_access": {
                        "has_access": true,
                        "error_code": null,
                        "developer_message": null,
                        "user_message": null,
                        "additional_context_user_message": null,
                        "user_fragment": null
                    },
                    "media": {
                        "course_image": {
                            "uri": "/asset-v1:a+a+a+type@asset+block@images_course_image.jpg",
                            "name": "Course Image"
                        }
                    },
                    "course_image": "/asset-v1:a+a+a+type@asset+block@images_course_image.jpg",
                    "course_about": "http://local.edly.io:8000/courses/course-v1:a+a+a/about",
                    "course_sharing_utm_parameters": {
                        "facebook": "utm_medium=social&utm_campaign=social-sharing-db&utm_source=facebook",
                        "twitter": "utm_medium=social&utm_campaign=social-sharing-db&utm_source=twitter"
                    },
                    "course_updates": "http://local.edly.io:8000/api/mobile/v4/course_info/course-v1:a+a+a/updates",
                    "course_handouts": "http://local.edly.io:8000/api/mobile/v4/course_info/course-v1:a+a+a/handouts",
                    "discussion_url": null,
                    "video_outline": null,
                    "is_self_paced": false
                },
                "certificate": {},
                "course_modes": [
                    {
                        "slug": "audit",
                        "sku": null,
                        "android_sku": null,
                        "ios_sku": null,
                        "min_price": 0
                    }
                ],
                "course_progress": {
                    "total_assignments_count": 0,
                    "assignments_completed": 0
                }
            }
        ]
    },
    "primary": {
        "audit_access_expires": null,
        "created": "2024-05-07T11:39:23.149565Z",
        "mode": "audit",
        "is_active": true,
        "course": {
            "id": "course-v1:b+b+b",
            "name": "b",
            "number": "b",
            "org": "b",
            "start": "2020-01-01T00:00:00Z",
            "start_display": "Jan. 1, 2020",
            "start_type": "timestamp",
            "end": null,
            "dynamic_upgrade_deadline": null,
            "subscription_id": "course_MNXXK4TTMUWXMMJ2MIVWEK3C",
            "courseware_access": {
                "has_access": true,
                "error_code": null,
                "developer_message": null,
                "user_message": null,
                "additional_context_user_message": null,
                "user_fragment": null
            },
            "media": {
                "course_image": {
                    "uri": "/asset-v1:b+b+b+type@asset+block@smile_742751.png",
                    "name": "Course Image"
                }
            },
            "course_image": "/asset-v1:b+b+b+type@asset+block@smile_742751.png",
            "course_about": "http://local.edly.io:8000/courses/course-v1:b+b+b/about",
            "course_sharing_utm_parameters": {
                "facebook": "utm_medium=social&utm_campaign=social-sharing-db&utm_source=facebook",
                "twitter": "utm_medium=social&utm_campaign=social-sharing-db&utm_source=twitter"
            },
            "course_updates": "http://local.edly.io:8000/api/mobile/v4/course_info/course-v1:b+b+b/updates",
            "course_handouts": "http://local.edly.io:8000/api/mobile/v4/course_info/course-v1:b+b+b/handouts",
            "discussion_url": null,
            "video_outline": null,
            "is_self_paced": false
        },
        "certificate": {},
        "course_modes": [
            {
                "slug": "audit",
                "sku": null,
                "android_sku": null,
                "ios_sku": null,
                "min_price": 0
            }
        ],
        "course_status": null,
        "course_progress": {
            "total_assignments_count": 10,
            "assignments_completed": 2
        },
        "course_assignments": {
            "future_assignments": [
                {
                    "assignment_type": "Homework",
                    "complete": false,
                    "date": "2024-05-31T00:00:00Z",
                    "date_type": "assignment-due-date",
                    "description": "",
                    "learner_has_access": true,
                    "link": "http://local.edly.io:8000/courses/course-v1:b+b+b/jump_to/block-v1:b+b+b+type@sequential+block@47eef9493e5b4d5b93a512d9f7751ba8",
                    "link_text": "",
                    "title": "Subsection 1-1",
                    "extra_info": null,
                    "first_component_block_id": "block-v1:b+b+b+type@problem+block@b4c5b0fef70643a581bd34471abf45b9"
                }
            ],
            "past_assignments": [
                {
                    "assignment_type": "exam 3",
                    "complete": false,
                    "date": "2024-04-24T00:00:00Z",
                    "date_type": "assignment-due-date",
                    "description": "",
                    "learner_has_access": true,
                    "link": "http://local.edly.io:8000/courses/course-v1:import+import+import/jump_to/block-v1:import+import+import+type@sequential+block@fee71303296a4a8cbc9542473ae94bd2",
                    "link_text": "",
                    "title": "Subsection 3-3",
                    "extra_info": null,
                    "first_component_block_id": "block-v1:import+import+import+type@problem+block@15dbe4ae8ca841bdb0f95ef86766d815"
                },
                {
                    "assignment_type": "Exam 1",
                    "complete": true,
                    "date": "2024-04-10T00:00:00Z",
                    "date_type": "assignment-due-date",
                    "description": "",
                    "learner_has_access": true,
                    "link": "http://local.edly.io:8000/courses/course-v1:import+import+import/jump_to/block-v1:import+import+import+type@sequential+block@82493e74c526482b9a077df1ebb32d2d",
                    "link_text": "",
                    "title": "Subsection",
                    "extra_info": null,
                    "first_component_block_id": "block-v1:import+import+import+type@problem+block@d7a5187c42454f99b96f468e0d9c1f1a"
                }
             ]
        }
    }
}
```
2. `BlocksInfoInCourseView` - we added to response 2 new fields:

- `assignment_progress`:

```
            "assignment_progress": {
                "assignment_type": "one",
                "num_points_earned": 2.0,
                "num_points_possible": 3.0
            }
```

- `course_progress `:
```
    "course_progress": {
        "total_assignments_count": 3,
        "assignments_completed": 2
    }
```
        
All new functionality was covered with unit tests.

## Supporting information

https://openedx.atlassian.net/wiki/spaces/COMM/pages/3935928321/FC-0047+-+Mobile+Development+Phase+3

## Testing instructions

`pytest lms/djangoapps/mobile_api/tests` - this command runs all tests related to mobile_api aplication. This command should be running in the LMS container.

Also, this API can be tested manual with Postman or another API client:
1. The request must be executed by an authorized user. `JwtAuthentication, BearerAuthentication, SessionAuthentication` can be used;
2. GET `{{LMS}}/api/mobile/<API_version>/users/<username>/course_enrollments/`;
3. Additional fields can be requested only with API_version == v4;
4. Also additional query params can be added to API request:
  - `?requested_fields=course_progress` - returns `course_progress` for each user's enrollment.
  - `?status=in_progress` - returns user's enrollments filtered by `status`. It can by next statuses:
      ```
    ALL = 'all'
    IN_PROGRESS = 'in_progress'
    COMPLETED = 'completed'
    EXPIRED = 'expired'
      ```
  - If `?status=in_progress` query param is used - the response contains only filtered user's enrollments by status. And doesn't contain `primary` enrollment object.


## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

Related mobile PRs:

- https://github.com/openedx/openedx-app-android/pull/308
- https://github.com/openedx/openedx-app-ios/pull/434

KyryloKireiev committed:
feat: [AXM-24] Update structure for course enrollments API (#2515)

* feat: [AXM-24] Update structure for course enrollments API

* style: [AXM-24] Improve code style

* fix: [AXM-24] Fix student's latest enrollment filter

KyryloKireiev committed:
feat: [AXM-47] Add course_status field to primary object (#2517)

NiedielnitsevIvan committed:
feat: [AXM-40] add courses progress to enrollment endpoint (#2519)

* fix: workaround for staticcollection introduced in e40a01c

* feat: [AXM-40] add courses progress to enrollment endpoint

* refactor: [AXM-40] add caching to improve performance

* refactor: [AXM-40] add progress only for primary course

* refactor: [AXM-40] refactor enrollment caching optimization

---------

Co-authored-by: Glib Glugovskiy <glib.glugovskiy@raccoongang.com>

NiedielnitsevIvan committed:
feat: [AXM-53] add assertions for primary course (#2522)

* feat: [AXM-53] add assertions for primary course

* test: [AXM-53] fix tests

* style: [AXM-53] change future_assignment default value to None

* refactor: [AXM-53] add some optimization for assignments collecting

NiedielnitsevIvan committed:
feat: [AXM-33] create enrollments filtering by course completion statuses (#2532)

* feat: [AXM-33] create enrollments filtering by course completion statuses

* test: [AXM-33] add tests for filtrations

* style: [AXM-33] fix pylint issues

NiedielnitsevIvan committed:
feat: [AXM-236] Add progress for other courses (#2536)

KyryloKireiev committed:
fix: [AXM-277] Change _get_last_visited_block_path_and_unit_name method implementation (#2540)

KyryloKireiev committed:
feat: [AXM-297] Add progress to assignments in BlocksInfoInCourseView API (#2546)

* feat: [AXM-297, AXM-310] Add progress to assignments and total course progress

* feat: [AXM-297] Add progress to assignments

* style: [AXM-297] Try to fix linters (add docstrings)

* refactor: [AXM-297] Add typing, refactor methods

KyryloKireiev committed:
feat: [AXM-288] Change response to represent Future assignments (#2550)

* feat: [AXM-288] Change response to represent Future assignments the same way as past assignments

* refactor: [AXM-288] Refactor get_course_assignments

Co-authored-by: monteri <36768631+monteri@users.noreply.github.com>

* refactor: [AXM-288] Refactor get_course_assignments method

---------

Co-authored-by: monteri <36768631+monteri@users.noreply.github.com>

KyryloKireiev committed:
feat: [AXM-287,310,331] Change course progress calculation logic (#2553)

* feat: [AXM-287,310,331] Change course progress calculation logic

* style: [AXM-287,310,331] Remove commented code

* fix: [AXM-287,310,331] Change course assignments gather logic

openedx-webhooks commented:
Thanks for the pull request, @KyryloKireiev! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

e0d commented:
I find the content of the API response to be confusing.  I'll spend some more time to try to grok it better, but I have a couple of questions immediately.

1. Why are the configs included in this response?
2. If I understand the way primary works, I don't like it :) I don't think we should be changing the shape of the response data for convenience.  Why wouldn't primary just be an attribute on the enrollment record?  

Can you explain a bit more about why we need the primary indicator in the first place?  I seems to combine to things, recency of registration and recency of activity, into one concept.  

Given that it is hiding some pretty specific business logic behind the API.  I'm not deeply familiar with the new mobile APIs, but this would seem to indicate we have a "backend-for-frontend" here, not really a general purpose API.  Is that right?

KyryloKireiev commented:
Hi Edward! This explains some of the things you asked about.

1. We took the `primary_course` out of the array so that pagination doesn’t apply to it. This way, we have access to the primary_course on every page. So, on each page, we have access to the attributes of the primary_course.
2. The `primary_course` has a different structure and this object goes through a different serializer compared to the enrollments in the array. I think this is wrong because we will have an object in the array that is not similar to the other objects. Serializer for `primary_course ` - `CourseEnrollmentSerializerModifiedForPrimary` contains expensive calculations. But this serializer used only API v4. That is, you can easily switch to another version of the API and not receive the `primary_course` object if there is no need for it.
3. Sorting all courses by the last visit is too expensive, so it seems that getting 1 course is much more efficient.
4. We have configs in the response, since we inherited from the previous version of this api. configs appeared in the API starting from version 2.

kdmccormick commented:
Hi @KyryloKireiev , I can review this soon. In the meantime, could you help me understand the background of this PR better? I see https://openedx.atlassian.net/wiki/spaces/COMM/pages/3935928321/FC-0047+-+Mobile+Development+Phase+3, but that page is mostly blank.

GlugovGrGlib commented:
Hi @kdmccormick thank you!
I see how this may be confusing when onboarding into context of FC-0047.
I believe the following references will allow you to understand the wider background.

1) As Michelle [commented on the FC page](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3935928321/FC-0047+-+Mobile+Development%2C+Phase+3?focusedCommentId=4271636491), the scope for this project consists of several mobile initiatives that are interdependent.

- Dates and calendar personalization
- Notifications
- Offline mode
- Navigation improvements

  This PR is focused only on **Navigation improvements** part. However, the https://github.com/openedx/edx-platform/pull/34859 is required for Dates and calendar personalization features.

2) Please refer to product document, outlining key requirements for **Navigation improvements**: https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3781754910/Project+-+Streamlined+Course+Navigation+Sequences+FC-0046. Additionally you can find sibling documents with product requirements for 3 other initiatives included into FC-0047 in the same space.

3) Also you can check these 4 mobile PRs to see the final results for implemented UI as well as demo of the functionality.
The code in the PRs below is built on top of the API changes introduced in this PR.
Android:
https://github.com/openedx/openedx-app-android/pull/308
https://github.com/openedx/openedx-app-android/pull/323
IOS:
https://github.com/openedx/openedx-app-ios/pull/434
https://github.com/openedx/openedx-app-ios/pull/446

GlugovGrGlib commented:
Hi @kdmccormick would you be able to take a look this week?
Also this PR https://github.com/openedx/edx-platform/pull/34859 can be considered as a follow up to mobile API extensions

GlugovGrGlib commented:
Hey @jawad-khan If you could provide a review from your side as well, it would be really helpful. Thanks!

kdmccormick commented:
@GlugovGrGlib thanks for the context. Yes, I should be able to review this week.

kdmccormick commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -141,6 +186,123 @@ class Meta:
+    @staticmethod
+    def _get_last_visited_block_path_and_unit_name(
+        block_id: str
+    ) -> Union[Tuple[None, None], Tuple[List['XBlock'], str]]:  # noqa: F821
```
Rather than `noqa` codes could you use the human-readable `pylint: ignore=...` ?

GlugovGrGlib replied:
Hi, sure we will be able to update the format of linter directives, but what do you think about switching to other linters, for example [ruff](https://github.com/openedx/edx-platform/issues/32174) in the future? 
Maybe it's worth discussing smooth transition from pylint only supported comments to flake8 like # noqa, that are supported by most python linters nowadays, so it might ease the migration?

kdmccormick replied:
That's an interesting point, and I think it could be worth discussing separately. Currently, though, we almost exclusively use pylint directives instead of noqa directives, so I'd like to stay consistent rather than starting a new pattern in this PR.

If we want to switch from pylint directives to noqa in the future, we could use a script to do it in bulk.

KyryloKireiev replied:
Yes, I see, I'm going to change `# noqa` to `pylint: ignore=...`

kdmccormick commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -124,6 +136,39 @@ def get_course_modes(self, obj):
+
+        return data
+
+    def calculate_progress(self, model: CourseEnrollment) -> Dict[str, int]:
```
Thank you for annotating types! Three notes:

1. The return type for this function should be `Dict[int, int]`
2. The annotations in this app are not checked in CI. If you would like them to be automatically checked, you could add the mobile_api or mobile_api/users to [the allowlist in mypy.ini](https://github.com/openedx/edx-platform/blob/master/mypy.ini#L8). If not, that's OK, just be aware that you'll need to check your annotations by hand.
3. Totally optional, no need to make this change: since the platform is on Py3.11 now, you can just use the lowercase builtin collection names now. No need to import them from typing. For example, `dict[int, int]`, `list[str]`. And instead of  `Union[str, None]` you can do `str | None`. They have the same meaning, the new ones are just easier to write.

GlugovGrGlib replied:
Thank you for notes! Regarding the third point, we implemented this a while ago in Python 3.8. I agree that it's definitely worth changing it to a more convenient format now since the platform is on py3.11.

KyryloKireiev replied:
I think we need to use `Dict[str, int]` typing here, because we have dictionary with the next structure:
```
{
            'total_assignments_count': 5,
            'assignments_completed': 10,
}
```
key -> str type
value -> int type

This is described in the following docs:
https://peps.python.org/pep-0484/#the-typing-module

kdmccormick replied:
Yup, you're totally right, my mistake

kdmccormick commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -141,6 +186,123 @@ class Meta:
+        """
+        return self.calculate_progress(model)
+
+    def get_course_assignments(self, model: CourseEnrollment) -> Dict[str, Optional[List[Dict[str, str]]]]:
```
I would like this logic to exist in a central platform Python API instead of the Mobile API. If the Learning MFE or Learner Dash MFE ever needs this same data, they should be able to call this function instead of reimplementing it. Could you move it? Maybe lms/djangoapps/courseware/courses.py is the right place?

In general, when implementing Mobile API features, try to keep the student/grading/progress/access logic in core applications, and then just use the mobile_api as a thin wrapper for presenting that information to the mobile apps. That way, both the Web and Mobile versions of the platform can have consistent behavior and we will have less business logic to test.

KyryloKireiev replied:
@kdmccormick I agree, it sounds logical. I moved `get_course_assignments` here: `lms/djangoapps/courseware/courses.py`

kdmccormick commented on lms/djangoapps/mobile_api/course_info/utils.py:
```diff
@@ -0,0 +1,43 @@
+log = logging.getLogger(__name__)
+
+
+def calculate_progress(
```
Same as my comment on `get_course_assignments`, this is core business logic that needs to be implemented in core APIs for consistency with the Web views.

KyryloKireiev replied:
And this `calculate_progress ` function I also move to `lms/djangoapps/courseware/courses.py`. Everything works correctly and I'm going to push these changes

kdmccormick replied:
Sounds good. Let me know when you're ready for another review pass.

GlugovGrGlib replied:
@kdmccormick the comments were addressed, It's ready for the second round now

kdmccormick replied:
@KyryloKireiev I see that you moved one definition of `calculate_progress` to lms/djangoapps/courseware/courses.py, thank you. However, there is this second definition of `calculate_progress` below, still in the mobile API.

Firstly, I think this function should be called something else, as it returns subsection grades, not progress.

Secondly, and more importantly, I am concerned whether this implementation matches the other ways in which we show a user's subsection grades in edx-platform. For example, I assume that the gradebook and the progress page on the Learning MFE both show subsection grades. Does those implementations perform the same steps that we do here (load the block structure, cache it, calculate grades, and then re-calculate from visible-only) ?  If so, why not use a shared API function? If not, why?

cc @jawad-khan

KyryloKireiev replied:
1. @kdmccormick I renamed and moved all `calculate_progress ` methods to `lms/djangoapps/courseware/courses.py`;
2. @GlugovGrGlib will be able to answer the second question better. This decision was made taking into account the further development of the mobile API.

GlugovGrGlib replied:
@kdmccormick that's true this approach for the assignments completions and progress calculation is not similar to the one on progress page in LMS. This was a product decision made in collaboration with @marcotuts 
Please refer to [FIGMA designs](https://www.figma.com/design/iZ56YMjbRMShCCDxqrqRrR/Open-edX-Mobile-App-All-Screens-v2.1?node-id=9177-44208&t=meiXCsfDbe7B10Eq-0) and to the [product requirements write up](https://openedx.atlassian.net/wiki/spaces/OEPM/pages/3781754910/Project+-+Streamlined+Course+Navigation+Sequences+FC-0046#Course-Home) for more info.

We discussed that now data about assignments on progress page, and course home in web differs from mobile course home but in the future iterations the same feture may be applied to the web course home as well.

KyryloKireiev committed:
refactor: [AXM-506] Move progress logic to core apps

KyryloKireiev committed:
style: [AXM-506] Replace noqa to pylint disable

jawad-khan commented:
@KyryloKireiev @GlugovGrGlib I have started reviewing this PR.
 Since this PR is now adding code in student and courseware app, I'll recommend you to get a review from TNL and Aurora squad too.

GlugovGrGlib commented:
Hey @jawad-khan, thank you, so much!
I don't think we must wait for a review from other 2U teams for this PR as the pieces of code being added to the courseware and student apps are utility functions that don't change the external API in any way and are used only in the mobile_api app. However, it will be good to inform someone from these teams in the Slack. Thanks!

KyryloKireiev committed:
refactor: [AXM-506] Refactor get_course_assignments func

KyryloKireiev committed:
style: [AXM-506] Remove unused import

jawad-khan commented on openedx/features/course_duration_limits/access.py:
```diff
@@ -81,7 +81,7 @@ def get_user_course_expiration_date(user, course):
         return None
 
-    enrollment = CourseEnrollment.get_enrollment(user, course.id)
+    enrollment = CourseEnrollment.get_enrollment(user, course.id) if not enrollment else enrollment
```
Another approach(optional).
enrollment = enrollment or CourseEnrollment.get_enrollment(user, course.id)

jawad-khan commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -97,7 +107,7 @@ def get_audit_access_expires(self, model):
         Returns expiration date for a course audit expiration, if any or null
         """
-        return get_user_course_expiration_date(model.user, model.course)
+        return get_user_course_expiration_date(model.user, model.course, model)
```
why there is a need to send another param when function can work with two params?

KyryloKireiev replied:
I think we can pass `CourseEnrollment`  object into `get_user_course_expiration_date ` function and avoid the additional query to the database here:
```enrollment = enrollment or CourseEnrollment.get_enrollment(user, course.id)```

jawad-khan commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -124,6 +134,17 @@ def get_course_modes(self, obj):
             for mode in course_modes
         ]
 
+    def to_representation(self, instance: CourseEnrollment) -> 'OrderedDict':  # lint-amnesty, pylint: disable=unused-variable, line-too-long
```
why don't we add course_progress as serializer field?

KyryloKireiev replied:
We didn't want to have a `course_progress` field for each user's enrollment by default. So, we have made a decision to add `course_progress` field into each user's enrollment only when `course_progress` is passed into `requested_fields` query param.  

So if we were to add this field `course_progress` as serializer field - we would have this field in every version (v0.5, v1, v2, v3, v4) of this API. We would receive something like this: "course_progress": null. We didn't want to change the response for older versions of the API, so we added this field as an additional one to  `to_representation` method. 

For `primary` user enrollment we added `course_progress` as serializer field. Because this is new functionality and we don't change old functionality.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -396,6 +517,8 @@ def paginator(self):
 
         if self._paginator is None and api_version == API_V3:
             self._paginator = DefaultPagination()
+        if self._paginator is None and api_version == API_V4:
```
Are we expecting a default paginator for v4? 
If no then why api_version == API_V isn't enough here?

KyryloKireiev replied:
We need to use something like this here:
`if self._paginator is None and api_version == API_V4:`
Because the method `paginator` is called 4 times when we requesting the API. And only the first time we don't have pagination object. Then we have pagination object:
`<lms.djangoapps.mobile_api.users.views.UserCourseEnrollmentsV4Pagination object at 0x7f57017f5950>`
And we can't change pagination object, we need only return pagination object for the next method calls.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -341,47 +362,147 @@ def get_serializer_class(self):
-            return list(mobile_available)
+            return mobile_available
+
+    def get_mobile_available_enrollments(self) -> List[Optional[CourseEnrollment]]:
```
This method also filters same organization check, can we add this in method name?
maybe:
get_same_org_mobile_available_enrollments

KyryloKireiev replied:
Yes, renamed

kdmccormick replied:
The return type should be simply `list[CourseEnrollment]` -- the list items will never be None.

KyryloKireiev replied:
@kdmccormick yes, you are right. It should be `list[CourseEnrollment]` return type. I lost this comment. I'll add this fix to the next push.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -341,47 +362,147 @@ def get_serializer_class(self):
             if check_course_expired(self.request.user, enrollment.course) == ACCESS_GRANTED
         )
 
+        if api_version == API_V4 and status not in EnrollmentStatuses.values():
```
can we pass this method not_duration_limited list to make it work faster?

KyryloKireiev replied:
@jawad-khan сan you please explain what you mean? I don’t quite understand what needs to be done?

jawad-khan replied:
Sorry, I added comment on the wrong line.
I was just suggesting why not we send mobile_available to get_primary_enrollment_by_latest_enrollment_or_progress in the following line.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -341,47 +362,147 @@ def get_serializer_class(self):
+
+        mobile_available_course_ids = [enrollment.course_id for enrollment in mobile_available]
+
+        latest_enrollment = self.queryset.filter(
```
Haven't we already filtered is_activer and username in queryset method?
will it work the same way if we don't apply username and is_active check here?

KyryloKireiev replied:
Yes, you are right. It was extra filtration. I removed these filtrations.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -341,47 +362,147 @@ def get_serializer_class(self):
+            course__id__in=mobile_available_course_ids,
+        ).order_by('-created').first()
+
+        if not latest_enrollment:
```
latest_enrollment code looks duplicated, we are doing same thing in queryset?
can we avoid this?

kdmccormick replied:
furthermore, could we centralize the latest_enrollment calculation in a Python API function outside of the mobile_api? For example, `common/djangoapps/student/api.py:get_last_active_enrollment`

KyryloKireiev replied:
But we don't filter user's enrollments by `mobile_available` in `queryset `. Should we do this filtration is `queryset` method?

jawad-khan replied:
We need to redesign our queryset and get_queryset approach. once its finalized we can apply mobile_available filter in get_queryset and use it everywhere else.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -341,47 +362,147 @@ def get_serializer_class(self):
 
-    def get_queryset(self):
+    @cached_property
+    def queryset(self):
```
You should be very careful when using a class level attribute.
Unless we have a strong reason of doing this I'll advise that we move instance or request level logic to get_queryset.

KyryloKireiev replied:
Yes, you are right. For the first time we try to put all logic into `get_queryset `, but we received more then 400 SQL queries for the one request (!). So, we need to use property method for received `queryset`. Caching helps maximize performance, so we need to `@cached_property` decorator to make this API faster. Now we have about 120 SQL queries for the one API request. This is the maximum number of SQL queries when requesting additional fields and a large number of the user's enrollments.

jawad-khan replied:
This [stackflow answer](https://stackoverflow.com/questions/19707237/use-get-queryset-method-or-set-queryset-variable/19707855#19707855) the best example of when to use queryset and get_queryset.
In our case we should move all request level logic to get_queryset and use caching there. queryset is already cached since it is evaluated only once.

jawad-khan commented on lms/djangoapps/mobile_api/users/views.py:
```diff
@@ -341,47 +362,147 @@ def get_serializer_class(self):
                 'enrollments': response.data
             }
+            if api_version == API_V4 and status not in EnrollmentStatuses.values():
+                if status in EnrollmentStatuses.values():
```
Doesn't this condition have a conflict with above condition?
We are here because we already know that status is not in EnrollmentStatuses.values()

KyryloKireiev replied:
Yes, this is actually not a necessary condition. I removed it.

Here the logic may be a little unclear in isolation from mobile applications. If the `status` parameter is passed to the query params, then the `primary` enrollment object field should not exist. Only enrollments filtered by status should be returns by this API.

jawad-khan commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -141,6 +162,101 @@ class Meta:
+        super().__init__(*args, **kwargs)
+        self.course = modulestore().get_course(self.instance.course.id)
+
+    def get_course_status(self, model: CourseEnrollment) -> Optional[Dict[str, List[str]]]:
```
Seems like most of this logic is already written in this [view](https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/mobile_api/users/views.py#L90).
Can we reuse this logic without duplicating it?

KyryloKireiev replied:
Yes, I see. This is exactly what we did initially. We tried to use methods from this view. We used functions from this method `_last_visited_block_path`. We also moved the repetitive logic into a separate function. But our QAs found out that the method `_get_course_info` returns the wrong path to the last visited block. So, we decided to create new method that returns correct path to the last visited block and don't use logic from `_last_visited_block_path` method.

Maybe some problems are in this function:
```
def get_current_child(xblock, min_depth=None, requested_child=None):
    """
    Get the xblock.position's display item of an xblock that has a position and
    children.  If xblock has no position or is out of bounds, return the first
    child with children of min_depth.
```
We may not understand how this function is supposed to work. But the path to last visited block that is built using this function is not correct

kdmccormick replied:
@KyryloKireiev Does `get_current_child` return the wrong block, or is the right block with the wrong path?

KyryloKireiev replied:
@kdmccormick `get_current_child` function is used in `UserCourseStatus._last_visited_block_path` method. This method `_last_visited_block_path` returns incorrect path to the last visited block. Maybe `get_current_child` function works correctly. But I think it is not correct to use it to find path to the last visited block.

jawad-khan commented:
I have added my first review, will deploy these changes and test it locally once you give feedback on this.
Also I have asked TNL squad on their slack channel to look at these changes.

kdmccormick reviewed (commented):
I agree with the requests @jawad-khan has raised and I've added a couple comments to his threads. Otherwise, my only other request is that under all PRs' "Testing Instructions" you describe which flags/data you used to manually smoke-test test the changes.

kdmccormick commented:
Regarding TNL/Aurora review-- if someone from that teams wants to jump in and review as well they are welcome to, but between you and I @jawad-khan, I think our approvals are enough to merge this.

kdmccormick commented on lms/djangoapps/mobile_api/users/serializers.py:
```diff
@@ -141,6 +162,101 @@ class Meta:
+            vertical = last_visited_block.get_parent()
+            sequential = vertical.get_parent()
+            chapter = sequential.get_parent()
+            course = chapter.get_parent()
```
Querying blocks from modulstore is very inefficient and should be [avoided whenever possible](https://github.com/openedx/edx-platform/blob/master/docs/decisions/0011-limit-modulestore-use-in-lms.rst). All you need from the vertical, sequential, chapter, and course are their usage keys, right? There must be a more efficient way to get those. Check out the block_structures API.

KyryloKireiev replied:
Yes, I rewrote this logic. Now the path to the last visited block is calculated in the same way as on the web. I used `path_to_location` function, this function used for the `Resume` button on web.

kdmccormick replied:
Looks great, thanks!

KyryloKireiev commented:
> I agree with the requests @jawad-khan has raised and I've added a couple comments to his threads. Otherwise, my only other request is that under all PRs' "Testing Instructions" you describe which flags/data you used to manually smoke-test test the changes.

@kdmccormick I added detail instruction how make manual API testing to the PR's description.

KyryloKireiev committed:
refactor: [AXM-506] Fix code review issues

KyryloKireiev commented:
Hi, @kdmccormick @jawad-khan I tried to fix the comments, but I also left a few questions. Please check it out when you have a chance

KyryloKireiev committed:
refactor: [AXM-506] Rename progress calculation functions

GlugovGrGlib commented on lms/djangoapps/courseware/courses.py:
```diff
@@ -764,6 +773,38 @@ def _ora_assessment_to_assignment(
+    is_staff = bool(has_access(user, 'staff', course_id))
+
+    try:
+        cache_key = f'course_block_structure_{str(course_id)}_{user.id}'
```
I believe we should use `course_version` attribute to make sure that cache is invalidated with each new published course version
Please see https://github.com/openedx/edx-platform/blob/master/lms/djangoapps/course_home_api/outline/views.py#L456 for similar caching approach.

KyryloKireiev replied:
Yes, added cache invalidation with course version

jawad-khan commented on common/djangoapps/student/models/course_enrollment.py:
```diff
@@ -129,11 +129,71 @@ class UnenrollmentNotAllowed(CourseEnrollmentException):
+        """
+        return self.filter(is_active=True)
+
+    def without_certificates(self, user_username):
```
username is enough.

jawad-khan commented on common/djangoapps/student/models/course_enrollment.py:
```diff
@@ -129,11 +129,71 @@ class UnenrollmentNotAllowed(CourseEnrollmentException):
+        from lms.djangoapps.certificates.models import GeneratedCertificate  # pylint: disable=import-outside-toplevel
+        course_ids_with_certificates = GeneratedCertificate.objects.filter(
+            user__username=user_username
+        ).values_list('course_id', flat=True)
```
These two lines are duplicated, can we write a method for user_course_ids_with_certificates?

jawad-khan commented:
@KyryloKireiev Once we agree on queryset implementation apparoach I'll test PR locally and will approve it.

KyryloKireiev committed:
refactor: [AXM-506] Refactor user's enrolments API

KyryloKireiev commented:
> @KyryloKireiev Once we agree on queryset implementation apparoach I'll test PR locally and will approve it.

Hi @jawad-khan, I pushed new `queryset` implementation. I thought a lot about how this could be changed. It seems to me that the only available way - create new property method `queryset_for_user` and call this method where we need. Now we didn't override `queryset` method. 

The main problem of the queryset implementation is `get_same_org_mobile_available_enrollments` and `mobile_available` filtration in this method. Because `mobile_available` has type `list` and we can't to do new filtrations with `mobile_available` list, but we need to do filtrations with queryset in `get_primary_enrollment_by_latest_enrollment_or_progress`. Unfortunately I don't see the possibility to get `mobile_available` enrollments in `get_queryset` method and with `QuerySet` type.  I pushed code with current changes.

@jawad-khan, if my approach seems wrong to you, can you explain in more detail how it should be?

jawad-khan commented:
> > @KyryloKireiev Once we agree on queryset implementation apparoach I'll test PR locally and will approve it.
> 
> Hi @jawad-khan, I pushed new `queryset` implementation. I thought a lot about how this could be changed. It seems to me that the only available way - create new property method `queryset_for_user` and call this method where we need. Now we didn't override `queryset` method.
> 
> The main problem of the queryset implementation is `get_same_org_mobile_available_enrollments` and `mobile_available` filtration in this method. Because `mobile_available` has type `list` and we can't to do new filtrations with `mobile_available` list, but we need to do filtrations with queryset in `get_primary_enrollment_by_latest_enrollment_or_progress`. Unfortunately I don't see the possibility to get `mobile_available` enrollments in `get_queryset` method and with `QuerySet` type. I pushed code with current changes.
> 
> @jawad-khan, if my approach seems wrong to you, can you explain in more detail how it should be?

This implementation can be another of not using class level attribute(which is recommended). I'll test it locally. Meanwhile can you please fix failing test cases?

KyryloKireiev committed:
fix: [AXM-506] Try to fix pipelines

KyryloKireiev commented:
@jawad-khan pipelines is passed successfully, ready to local review

jawad-khan commented:
@kdmccormick I have approved these changes and its clear from my end, can you please finalize your review because we need theses changes asap.

GlugovGrGlib commented:
Hi @kdmccormick, would you be able to do another round of review?

kdmccormick commented:
I am out on holiday and unable to review until 16 Jul, but I do trust @jawad-khan 's review so don't hesistate to merge if this is ready to go.

GlugovGrGlib commented:
@kdmccormick Ok, thanks for the update, have a nice holiday!

@jawad-khan Let's merge it as soon as you're ready?

jawad-khan commented:
@kdmccormick can you please resolve your concerns if they are addressed?

jristau1984 commented:
I have checked over the `student/models/` file changes, and they seem acceptable from the SOX perspective.

miankhalid commented:
was trying to see if removing Kyle's review would enable the merge button 👆

openedx-webhooks commented:
@KyryloKireiev 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34848&entry.1484655318=https://github.com/KyryloKireiev&entry.752974735=2024-05-23+16:34&entry.1917517419=2024-07-10+15:07&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->

hassan6190 commented:
@GlugovGrGlib Can you help me understand how does this progress bar track progress? Are there any specific xblock completion it is tracking?
